rfd/0129-discovery-name-templating.md

```diff
@@ -0,0 +1,241 @@
+---
+authors: Gavin Frazar (gavin.frazar@goteleport.com)
+state: draft
+---
+
+# RFD 0129 - Discovery Resource Name Templates
+
+Related RFDs:
+- [RFD 0125 - Dynamic Auto-Discovery Configuration](./0125-dynamic-auto-discovery-config.md)
+
+## Required Approvers
+
+- Engineering: `@r0mant && @smallinsky && @tigrato`
+- Product: `@klizhentas || @xinding33`
+- Security: `@reedloden || @jentfoo`
+
+## What
+
+Auto-Discovery service can be configured to dynamically name discovered
+resources using a template string that can reference a subset of resource
+metadata. The choice of what metadata to make available is the main motivation
+for this RFD.
+
+## Why
+
+Multiple discovery agents can discover resources with identical names.
+For example, this happened when customers had databases in different AWS
+regions or accounts with the same name. When a name collision occurs, only one
+of the databases can be accessed by users.
+
+Resource name templates can be used to avoid naming collisions by renaming
+discovered resources with a prefix/suffix specific to a discovery agent, or by
+using other available metadata such as AWS region or account ID.
+
+See:
+- https://github.com/gravitational/teleport/issues/22438
+
+## Details
+
+There will be a new optional configuration string in Discovery service matchers,
+`resource_name_template`, that will be parsed as a
+[Go Template](https://pkg.go.dev/text/template) to rename resources discovered
+using that matcher.
+
+This setting will be made available in both static config and the upcoming
+dynamic discovery config from RFD 125.
+
+### `resource_name_template`
+
+This template string is just a Go text/template, so the template syntax is the
+same.
+
+Using Go templates is not uncommon, so this UX should not be surprising.
+A few examples of software that supports using Go templates: `helm`, `gh`
+(GitHub CLI), `docker`.
+
+### Example config:
```

Comment:
I have a couple of questions before we proceed:

* How would cloud users configure this?
* Can users add and remove tags with this?
* Can we re-use automation we already built for login rules (see the login rules RFD).

cc @r0mant

Reply:
> How would cloud users configure this?

With `tctl create`, using the new dynamic resource `DiscoveryConfig` from [RFD 125](https://github.com/gravitational/teleport/blob/479accbed79fd036bbc0c8979c20942fb787c994/rfd/0125-dynamic-auto-discovery-config.md) (not yet implemented)

> Can users add and remove tags with this?

No, it's just for resource name rewriting in Teleport.

Would tag editing be useful for preventing discovered resource name collisions?

Or do you have a different use case in mind for tag editing?

> Can we re-use automation we already built for login rules (see the login rules RFD).

I'm not very familiar with what Teleport predicate lang is capable of - I'll review that and get back to this question, but from skimming the RFD, it seems to be focused on string set/dict manipulation rather than string manipulation.

Since this is manipulating a single string (the Teleport dynamic resource's name), only a subset of predicate lang would be supported.

Predicate lang would need to support more complex string manipulation,
e.g. `"fooDB" -> "fooDB-<region>-<accountID>"`, which we could add if it's not already in there I suppose.

Reply:
@GavinFrazar yes, would like to see more unity with [login rules](https://github.com/gravitational/teleport/blob/479accbed79fd036bbc0c8979c20942fb787c994/rfd/0078-login-rules.md) and add ability to not only rename resource names by adding suffixes, but also add remove tags, because many customers expressed interest in this feature, especially removing extra tags.

Please add examples using new syntax for dynamic configuration in your RFD. Cloud becomes primary use-case for us, so it's no longer enough to add static config files examples only.

Reply:
The syntax for this `resource_name_template` setting will be identical between static config and the dynamic `DiscoveryConfig` - just a string field in the aws/gcp/azure matchers

Reply:
@klizhentas just to be sure, when you say "tags" you're referring to tags in the cloud i.e. AWS resource tags, not Teleport labels?

Reply:
yes, converting AWS tags to teleport labels, and dropping/renaming some tags during conversion

Reply:
marking this resolved since we decided against templating